nexus/src/external_api/shared.rs

```diff
@@ -40,12 +40,56 @@ pub const MAX_ROLE_ASSIGNMENTS_PER_RESOURCE: usize = 64;
 /// access to this resource.
 #[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize, JsonSchema)]
 #[schemars(rename = "{AllowedRoles}Policy")]
-pub struct Policy<AllowedRoles: serde::de::DeserializeOwned> {
+pub struct Policy<AllowedRoles: serde::de::DeserializeOwned + std::clone::Clone>
+{
     /// Roles directly assigned on this resource
     #[serde(deserialize_with = "role_assignments_deserialize")]
     pub role_assignments: Vec<RoleAssignment<AllowedRoles>>,
 }
 
+impl<AllowedRoles: serde::de::DeserializeOwned + std::clone::Clone>
+    Policy<AllowedRoles>
+{
+    pub fn add_or_update_role_assignment(
```

Comment:
Is this only used for the case where we're assigning a Silo rule during JIT provision?  In that case we know the user has no roles.  I think it would be a lot simpler to insert a new role assignment (which you could do directly in the caller)